File: app/src/main/java/im/tox/antox/utils/AntoxFriendList.java
package im.tox.antox.utils;

import im.tox.jtoxcore.FriendExistsException;
import im.tox.jtoxcore.FriendList;
import im.tox.jtoxcore.ToxFriend;
import im.tox.jtoxcore.ToxUserStatus;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;
import java.util.Locale;

/**
 * Wrapper class for the Tox Friendlist.
 * 
 * @author sonOfRa
 * 
 */
public class AntoxFriendList implements FriendList<AntoxFriend> {

	/**
	 * Underlying friend list
	 */
	private List<AntoxFriend> friends;

	/**
	 * Create a new, empty ToxFriendList instance
	 */
	public AntoxFriendList() {
		this.friends = Collections
				.synchronizedList(new ArrayList<AntoxFriend>());
	}

	/**
	 * Create a new ToxFriendList instance with the given List of Friends as an
	 * underlying structure
	 * 
	 * @param friends
	 *            initial list of friends to use
	 */
	public AntoxFriendList(ArrayList<AntoxFriend> friends) {
		this.friends = friends;
	}

	@Override
	public AntoxFriend getByFriendNumber(int friendnumber) {
		synchronized (this.friends) {
			for (AntoxFriend friend : this.friends) {
				if (friend.getFriendnumber() == friendnumber) {
					return friend;
				}
			}
		}
		return null;
	}

	@Override
	public AntoxFriend getById(String id) {
        if(id!=null) {
            synchronized (this.friends) {
                for (AntoxFriend friend : this.friends) {
                    if (id.equals(friend.getId())) {
                        return friend;
                    }
                }
            }
        }
		return null;
	}

	@Override
	public List<AntoxFriend> getByName(String name, boolean ignorecase) {
		if (ignorecase) {
			return getByNameIgnoreCase(name);
		}

		ArrayList<AntoxFriend> result = new ArrayList<AntoxFriend>();
		synchronized (this.friends) {
			for (AntoxFriend f : this.friends) {
				if (name == null && f.getName() == null) {
					result.add(f);
				} else if (name != null && name.equals(f.getName())) {
					result.add(f);
				}
			}
		}
		return result;
	}

	private List<AntoxFriend> getByNameIgnoreCase(String name) {
		ArrayList<AntoxFriend> result = new ArrayList<AntoxFriend>();
		synchronized (this.friends) {
			for (AntoxFriend f : this.friends) {
				if (name == null && f.getName() == null) {
					result.add(f);
				} else if (name != null && name.equalsIgnoreCase(f.getName())) {
					result.add(f);
				}
			}
		}
		return result;
	}

	@Override
	public List<AntoxFriend> searchFriend(String partial) {
		if (partial == null) {
			throw new IllegalArgumentException("Cannot search for null");
		}
		String partialLowered = partial.toLowerCase(Locale.US);
		ArrayList<AntoxFriend> result = new ArrayList<AntoxFriend>();
		synchronized (this.friends) {
			for (AntoxFriend f : this.friends) {
				String name = (f.getName() == null) ? null : f.getName()
						.toLowerCase(Locale.US);
				if (name.contains(partialLowered)) {
					result.add(f);
				}
			}
		}
		return result;
	}

	@Override
	public List<AntoxFriend> getByStatus(ToxUserStatus status) {
		ArrayList<AntoxFriend> result = new ArrayList<AntoxFriend>();
		synchronized (this.friends) {
			for (AntoxFriend f : this.friends) {
				if (f.isOnline() && f.getStatus() == status) {
					result.add(f);
				}
			}
		}
		return result;
	}

	@Override
	public List<AntoxFriend> getOnlineFriends() {
		ArrayList<AntoxFriend> result = new ArrayList<AntoxFriend>();
		synchronized (this.friends) {
			for (AntoxFriend f : this.friends) {
				if (f.isOnline()) {
					result.add(f);
				}
			}
		}
		return result;
	}

	@Override
	public List<AntoxFriend> getOfflineFriends() {
		ArrayList<AntoxFriend> result = new ArrayList<AntoxFriend>();
		synchronized (this.friends) {
			for (AntoxFriend f : this.friends) {
				if (!f.isOnline()) {
					result.add(f);
				}
			}
		}
		return result;
	}

	@Override
	public List<AntoxFriend> all() {
		return new ArrayList<AntoxFriend>(this.friends);
	}

	@Override
	public AntoxFriend addFriend(int friendnumber) throws FriendExistsException {
		AntoxFriend friend;
		synchronized (this.friends) {
			for (AntoxFriend f : this.friends) {
				if (f.getFriendnumber() == friendnumber) {
					throw new FriendExistsException(f.getFriendnumber());
				}
			}
			friend = new AntoxFriend(friendnumber);
			this.friends.add(friend);
		}
		return friend;
	}

	@Override
	public AntoxFriend addFriendIfNotExists(int friendnumber) {
		synchronized (this.friends) {
			for (AntoxFriend f : this.friends) {
				if (f.getFriendnumber() == friendnumber) {
					return f;
				}
			}
			AntoxFriend friend = new AntoxFriend(friendnumber);
			this.friends.add(friend);
			return friend;
		}
	}

	@Override
	public void removeFriend(int friendnumber) {
		synchronized (this.friends) {
			Iterator<AntoxFriend> it = this.friends.iterator();
			while (it.hasNext()) {
				ToxFriend f = it.next();
				if (f.getFriendnumber() == friendnumber) {
					it.remove();
					break;
				}
			}
		}
	}

}

File: app/src/main/scala/im/tox/antox/utils/AntoxFriendList.scala
package im.tox.antox.utils

import java.util
import java.util.{ArrayList, Collections, List, Locale}

import im.tox.antox.tox.ToxSingleton
import im.tox.tox4j.core.enums.ToxStatus
import im.tox.tox4j.core.exceptions.ToxFriendAddException
import im.tox.tox4j.exceptions.ToxException

import scala.collection.JavaConversions._

class AntoxFriendList {

  private var friends: util.List[AntoxFriend] = Collections.synchronizedList(new util.ArrayList[AntoxFriend]())

  def this(friends: util.ArrayList[AntoxFriend]) {
    this()
    this.friends = friends
  }

  def getByFriendNumber(friendNumber: Int): Option[AntoxFriend] = {
    friends.find(friend => friend.getFriendnumber == friendNumber)
  }

  def getByKey(key: String): Option[AntoxFriend] = {
    friends.find(friend => friend.getKey == key)
  }

  def getByName(name: String, ignorecase: Boolean): util.List[AntoxFriend] = {
    if (ignorecase) {
      return getByNameIgnoreCase(name)
    } else {
      friends.filter(friend => (friend.name == null && name == null) || (name != null && name == friend.name))
    }
  }

  private def getByNameIgnoreCase(name: String): util.List[AntoxFriend] = {
    friends.filter(friend => (friend.name == null && name == null) || (name != null && name.equalsIgnoreCase(friend.name)))
  }

  def searchFriend(partial: String): util.List[AntoxFriend] = {
    val partialLowered = partial.toLowerCase(Locale.US)
    if (partial == null) {
      throw new IllegalArgumentException("Cannot search for null")
    }
    friends.filter(friend => (friend.name != null && friend.name.contains(partialLowered)))
  }

  def getByStatus(status: ToxStatus): util.List[AntoxFriend] = {
    friends.filter(friend => friend.isOnline && friend.getStatus == status)
  }

  def getOnlineFriends: util.List[AntoxFriend] = {
    friends.filter(friend => friend.isOnline)
  }

  def getOfflineFriends: util.List[AntoxFriend] = {
    friends.filter(friend => !friend.isOnline)
  }

  def all(): util.List[AntoxFriend] = {
    new util.ArrayList[AntoxFriend](this.friends)
  }

  def addFriend(friendnumber: Int): AntoxFriend = {
    friends.filter(friend => friend.getFriendnumber == friendnumber).headOption match {
      case Some(f) => throw new Exception()
      case None => {
        val f = new AntoxFriend(friendnumber)
        this.friends.add(f)
        f
      }
    }
  }

  def addFriendIfNotExists(friendnumber: Int): AntoxFriend = {
    friends.filter(friend => friend.getFriendnumber == friendnumber).headOption match {
      case Some(f) => f
      case None => {
        val f = new AntoxFriend(friendnumber)
        this.friends.add(f)
        f
      }
    }
  }

  def updateFromFriend(friend: Friend): Unit = {
    val antoxFriend = getByKey(friend.key).get
    antoxFriend.setName(friend.name)
    antoxFriend.setStatusMessage(friend.status)
    antoxFriend.setOnline(friend.isOnline)
  }

  def removeFriend(friendnumber: Int) {
    friends.remove(friends.find(friend => friend.getFriendnumber == friendnumber).get)
  }
}
